### PR TITLE
Issue 683

### PR DESCRIPTION
**Pull request by CharliePoole**
**Ready to Merge**

As explained in #683, we need to have FrameworkController working in the portable build. This PR does that and in fact incorporates FrameworkController in all builds, including CF and Silverlight. The class is small, so it's no skin lost if we don't actually use it and it's better to keep it working for future use.

This change incorporates the following:

* The NUnit.Framework.Interfaces.XmlNode class has been renamed to TNode to avoid confusion with the System.XmlNode class. It is entirely independent of platform except for parsing of XML. The parsing is done using XmlDocument and XmlNode where available and XElement otherwise. This is not a general-use XML class. It only supports what NUnit actually needs.

* TestFilter has been rewritten to use TNode rather than XMLNode.

* Changes have been made to FrameworkController to allow it to be used in Silverlight, CF and the portable build. Most notably, it uses the assembly display name to open assemblies under Silverlight.

* Various platform differences have been encapsulated in classes in the Compatibility directory.

The initial commit should be reviewed for any changes needed. Before merging, we ought to consider a few things that have not yet been done but which we may want to include:

1. Should we add an entry point to FrameworkController that takes an Assembly rather than a path or AssemblyNamel?

2. Should NUnitLite be modified to use FrameworkController rather than directly referencing the framework for which it is built and using the ITestAssemblyBuilder interface?

### CODE REVIEW

**Comment by oznetmaster** [comment]
I am getting every posting to this group emailed to me twice.  It is very annoying :(

**Comment by CharliePoole** [comment]
@nunit/core-team I'm blocked on this and on the portable engine till I get some review and comment.

File: src/NUnitFramework/framework/Interfaces/TNode.cs [hunk 1]
```diff
@@ -22,54 +22,44 @@
...
-    /// XmlNode represents a single node in the XML representation
+    /// TNode represents a single node in the XML representation
     /// of a Test or TestResult. It replaces System.Xml.XmlNode and
-    /// provides a minimal set of methods for operating on the XML 
-    /// in a platform-independent manner.
+    /// System.Xml.Linq.XElement, providing a minimal set of methods 
+    /// for operating on the XML in a platform-independent manner.
     /// </summary>
-    public class XmlNode
+    public class TNode
```
Comment (rprouse) [inline src/NUnitFramework/framework/Interfaces/TNode.cs]: Why TNode for the name? I am probably being overly picky, but when I read `TNode` in `IXmlNodeBuilder` I expected it to be a generic method.

Comment (CharliePoole) [inline src/NUnitFramework/framework/Interfaces/TNode.cs]: I modelled it after XNode and thought of it as a T(est)Node. I'm not 100% happy with the name but I can't think of anything better that won't conflict with the System.Xml or System.Xml.Linq namespeaces. Suggestions?

Comment (rprouse) [inline src/NUnitFramework/framework/Interfaces/TNode.cs]: The whole XML namespace is way too polluted with different classes for working with XML, so it doesn't leave much does it. NNode (Nunit) doesn't read well either. The only thing that I can think of is going longer like NunitXmlNode, but that doesn't really flow off the tongue either does it. We should probably leave it as is unless one of us suddenly has an epiphany :smile:

**Comment by rprouse** [comment]
The code looks good to me. For your two questions, I do think we should create an entry point that takes an assembly to be tested if it isn't too hard to plumb through. It allows us to run tests in the CurrentlyExecutingAssembly very easily, or in an assembly that contains a type. I also think we might run into problems on some platforms if we do not provide it.

As for modifying NUnitLite, I like the idea because it will help us exercise this interface and because other lite runners like Xamarin will be using it. It also allows us to decouple the release of the NUnitLite runners from the framework. I think this should probably be a second PR from this though.

**Comment by CharliePoole** [comment]
I'll do the assembly entry point as part of this PR. It should be simple enough. I agree that the NUnitLite change should be separate.

**Comment by CharliePoole** [comment]
Latest commit adds the new constructor. Ready to merge when CI passes.

**Pull request merged**